include/multipass/cli/format_utils.h

```diff
@@ -33,44 +34,77 @@ class Formatter;
 
 namespace format
 {
+static constexpr int col_buffer = 3;
+
 std::string status_string_for(const InstanceStatus& status);
 std::string image_string_for(const multipass::FindReply_AliasInfo& alias);
 Formatter* formatter_for(const std::string& format);
 
-template <typename Instances>
-Instances sorted(const Instances& instances);
+template <typename Container>
+Container sorted(const Container& items);
 
 void filter_aliases(google::protobuf::RepeatedPtrField<multipass::FindReply_AliasInfo>& aliases);
 
 // Computes the column width needed to display all the elements of a range [begin, end). get_width is a function
 // which takes as input the element in the range and returns its width in columns.
-static constexpr auto column_width = [](const auto begin, const auto end, const auto get_width, int minimum_width,
-                                        int space = 2) {
-    if (0 == std::distance(begin, end))
-        return minimum_width;
-
-    auto max_width =
-        std::max_element(begin, end, [&get_width](auto& lhs, auto& rhs) { return get_width(lhs) < get_width(rhs); });
-    return std::max(get_width(*max_width) + space, minimum_width);
-};
+static constexpr auto column_width =
+    [](const auto begin, const auto end, const auto get_width, int header_width, int minimum_width = 0) {
+        if (0 == std::distance(begin, end))
+            return std::max({header_width + col_buffer, minimum_width});
+
+        auto max_width = std::max_element(begin, end, [&get_width](auto& lhs, auto& rhs) {
+            return get_width(lhs) < get_width(rhs);
+        });
+        return std::max({get_width(*max_width) + col_buffer, header_width + col_buffer, minimum_width});
+    };
 } // namespace format
 } // namespace multipass
 
-template <typename Instances>
-Instances multipass::format::sorted(const Instances& instances)
+template <typename Container>
+Container multipass::format::sorted(const Container& items)
 {
-    if (instances.empty())
-        return instances;
+    if (items.empty())
+        return items;
 
-    auto ret = instances;
+    auto ret = items;
     const auto petenv_name = MP_SETTINGS.get(petenv_key).toStdString();
     std::sort(std::begin(ret), std::end(ret), [&petenv_name](const auto& a, const auto& b) {
-        if (a.name() == petenv_name)
+        using T = std::decay_t<decltype(a)>;
+        using google::protobuf::util::TimeUtil;
```

Comment:
@sharder996 credit is yours on this one, so feel free to reply, you're probably better able to explain.

My understanding is that yes, the main reason behind this is minimizing conversions necessary for communicating the timestamp between daemon and client. At the level of the formatters, the code is concerned with dealing with RPC replies, so I thought it was OK.

As for the nanosecond resolution, we need more resolution than just seconds. Otherwise, the [ordering would be messed up when snapshots were taken in quick succession](https://github.com/canonical/multipass/pull/3125#discussion_r1245092803). Miliseconds might be enough, but we'd be relying on no two snapshots being taken in the same millisecond, which wouldn't fully account for parallel requests. But it might be just that it is just that the type doesn't report milliseconds, only nanos.

Note that we show only milliseconds to the user, not nanoseconds. And we do so as part of a standard format, IIRC. This may be affected by the user locale PR, I haven't checked yet (@sharder996?).

Reply:
Ok, so the answer to this and the below question revolves around using the `google.protobuf.Timestamp` type in the protobuf message.

I didn't look at using the Poco library, but looked at using stl vs Qt. Qt just offered some useful functions for marshalling the date to a readable string for storing in json format. Having a look again at it now, I think the same functionality is possible with just stl.

The perceived nanosecond resolution comes from the fact that `google.protobuf.Timestamp` has only two functions for setting time; `set_seconds()` and `set_nanos()`. The initial time is captured in milliseconds and serialized in milliseconds (which is all that `QDateTime` supports). So, `TimeUtil::TimestampToNanoseconds` probably came from `set_nanos()` in `Timestamp`.

We only show millisecond resolution to the user, and internally I think it should be enough to prevent time conflicts.

Reply:
Ok, I think this is fine the way it is.  Thanks for the explanations @ricab and @sharder996!